packages/webview_flutter/webview_flutter_wkwebview/lib/src/common/instance_manager.dart

```diff
@@ -2,51 +2,204 @@
+    return identifier;
+  }
 
-  /// Attempt to add a new instance.
+  /// Removes the instance, if present, and call [onWeakReferenceRemoved] with
+  /// its identifier.
+  ///
+  /// Returns the identifier associated with the removed instance. Otherwise,
+  /// `null` if the instance was not found in this manager.
   ///
-  /// Returns new if [instance] has already been added. Otherwise, it is added
-  /// with a new instance id.
-  int? tryAddInstance(Object instance) {
-    if (_instancesToInstanceIds.containsKey(instance)) {
+  /// This does not remove the the strong referenced instance associated with
+  /// [instance]. This can be done with [removeStrongReference].
+  int? removeWeakReference(Copyable instance) {
+    final int? identifier = getIdentifier(instance);
+    if (identifier == null) {
       return null;
     }
 
-    final int instanceId = _nextInstanceId++;
-    _instancesToInstanceIds[instance] = instanceId;
-    _instanceIdsToInstances[instanceId] = instance;
-    return instanceId;
+    _identifiers[instance] = null;
+    //_finalizer.detach(instance);
+    onWeakReferenceRemoved(identifier);
+
+    return identifier;
+  }
+
+  /// Removes [identifier] and its associated strongly referenced instance, if
+  /// present, from the manager.
+  ///
+  /// Returns the strong referenced instance associated with [identifier] before
+  /// it was removed. Returns `null` if [identifier] was not associated with
+  /// any strong reference.
+  ///
+  /// This does not remove the the weak referenced instance associtated with
+  /// [identifier]. This can be done with [removeWeakReference].
+  Copyable? removeStrongReference(int identifier) {
+    return _strongInstances.remove(identifier);
   }
 
-  /// Remove the instance from the manager.
+  /// Retrieve the instance associated with identifier.
   ///
-  /// Returns null if the instance is removed. Otherwise, return the instanceId
-  /// of the removed instance.
-  int? removeInstance<T extends Object>(T instance) {
-    final int? instanceId = _instancesToInstanceIds[instance];
-    if (instanceId != null) {
-      _instancesToInstanceIds.remove(instance);
-      _instanceIdsToInstances.remove(instanceId);
+  /// The value returned is chosen in this order:
+  /// 1. A weakly referenced instance asscociated with identifier.
+  /// 2. When [returnedInstanceMayBeUsed] is set to `true` and the only instance
+  /// associated with identifier is a strongly referenced instance, a copy of the
+  /// instance is added as a weakly reference with the same instance id. Returning
+  /// the newly created weakly referenced copy.
+  /// 3. When [returnedInstanceMayBeUsed] is set to `false` and the only
+  /// instance associated with identifier has a strong reference. The strongly
+  /// referenced instance is returned.
+  /// 4. If no instance is associated with identifier, returns null.
+  T? getInstance<T extends Copyable>(
+    int identifier, {
+    required bool returnedInstanceMayBeUsed,
```

Comment:
Even after reading the comment, I'm not sure what "may be used" means conceptually. What would it mean to return an object that can't be used? Isn't it being used by definition if something is calling this method to look it up?